### PR TITLE
feat(LENS-1390): add support for incremental catalog updates in Java SDK

### PR DESCRIPTION
**Pull request by jgwatkincoveo**
Added the capability to use the stream mechanism in the Push API to enable incremental updates as per https://docs.coveo.com/en/l62e0540

### CODE REVIEW

**Review by louis-bompart — APPROVED** [review]
Cursive review, I mostly 👀 the public APIs for consistency sake for the users.
I leave the details of the implementation to y'all.

LGTM, double check the bump v with @y-lakhdar before merging

File: src/main/java/com/coveo/pushapiclient/StreamOperationType.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.coveo.pushapiclient;
+
+public enum StreamOperationType {
+  REBUILD("rebuild"),
+  INCREMENTAL("incremental");
```
Comment (samisayegh) [inline src/main/java/com/coveo/pushapiclient/StreamOperationType.java:5]: Could it be clearer if the names here aligned with [the docs](https://docs.coveo.com/en/l62e0540/coveo-for-commerce/how-to-update-your-catalog#leading-practices)?
`Rebuild` is used in brackets [here](https://docs.coveo.com/en/lb4a0344/coveo-for-commerce/stream-your-catalog-data-to-your-source#:~:text=during%20the%20stream-,(rebuild),-process%20if%20all), but is `incremental` referenced?

I see references "full item updates" and "partial item updates" so maybe it's more consistent to use such names?

File: src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java [hunk 2]
```diff
@@ -8,52 +9,120 @@
...
+      this.streamId = this.openStream();
     }
     queue.add(document);
     return this.streamId;
   }
 
+  public FileContainer addOrUpdate(DocumentBuilder document)
+      throws IOException, InterruptedException {
+    if (StreamOperationType.REBUILD.equals(getOperationType())) {
+      throw new UnsupportedOperationException(
```
Comment (samisayegh) [inline src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java:48]: Seeing this error, would it be cleaner if the `addOrUpdate` and `delete` methods were part of a dedicated class that could only be instantiated with the supported stream type?

File: src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java [hunk 3]
```diff
@@ -8,52 +9,120 @@
...
-      this.streamId = this.getStreamId();
+      this.streamId = this.openStream();
     }
     queue.add(document);
     return this.streamId;
   }
 
+  public FileContainer addOrUpdate(DocumentBuilder document)
+      throws IOException, InterruptedException {
+    if (StreamOperationType.REBUILD.equals(getOperationType())) {
```
Comment (samisayegh) [inline src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java:47]: Perhaps it's more resilient if the logic was inverted, for when other operation types are added in the future like "partial item updates"? e.g.

```
if (!StreamOperationType.INCREMENTAL.equals(getOperationType())) {...}
```

File: src/test/java/com/coveo/pushapiclient/StreamServiceInternalTest.java [hunk 4]
```diff
@@ -87,17 +107,132 @@ public void testCloseShouldFlushBufferedDocuments()
...
+          InterruptedException,
+          NoOpenStreamException,
+          NoOpenFileContainerException {
+    doReturn(StreamOperationType.REBUILD).when(service).getOperationType();
+    service.delete(deleteDocumentA);
+    service.close();
+  }
+
+  @Test(expected = UnsupportedOperationException.class)
+  public void testShouldThrowException_OnAdd_WhenRebuild()
```
Comment (samisayegh) [inline src/test/java/com/coveo/pushapiclient/StreamServiceInternalTest.java:220]: Should this test case be `whenIncremental`?

**Review by samisayegh — APPROVED** [review]
Great ✅

File: pom.xml [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
     <groupId>com.coveo</groupId>
     <artifactId>push-api-client.java</artifactId>
-    <version>2.4.0</version>
+    <version>2.5.0</version>
```
[the remark this answers is not in the document]
Comment (jgwatkincoveo) [reply pom.xml]: Done

File: src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java (hunk 2 above)
Comment (jgwatkincoveo) [reply src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java]: I originally took this approach to keep the stream concept in a single location, but it perhaps overcomplicated things. I refactored to use a separate internal service now, thereby invalidating some of the other comments.

File: src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java [hunk 6]
```diff
@@ -8,52 +9,120 @@
...
+      throw new UnsupportedOperationException(
+          "The full rebuild stream operation is not supported for the incramental addOrUpdate");
```
[the remark this answers is not in the document]
Comment (jgwatkincoveo) [reply src/main/java/com/coveo/pushapiclient/StreamServiceInternal.java]: No longer relevant based on separation of internal services

**Comment by jgwatkincoveo** [comment]
@y-lakdar Can you merge this for me please as I'm not authorised to. Thanks

**Comment by jgwatkincoveo** [comment]
@y-lakhdar can you merge this in for me please as I'm not authorised to. Thanks.

**Pull request merged**